test_conformance/subgroups/subhelpers.h

```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
-};
-template <> struct TypeCheck<cl_float>
+    lhs = static_cast<Ty>(rhs);
+}
+
+// set scalar value to half scalar
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value>::type
+set_value(Ty &lhs, const cl_ulong &rhs)
 {
-    static bool val(cl_device_id) { return true; }
-};
-template <> struct TypeCheck<cl_half>
+    lhs.data = rhs;
+}
+
+
+// compare for common vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
     {
-        return is_extension_available(device, "cl_khr_fp16");
+        // log_info("Values index i = %d got %lu expected %lu\n", i, lhs.s[i],
+        //         rhs.s[i]);
+        if (lhs.s[i] != rhs.s[i])
+        {
+            return false;
+        }
     }
-};
-template <> struct TypeCheck<double>
+    return true;
+}
+
+
+// compare for vectors 3
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_size3::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    for (auto i = 0; i < 3; ++i)
     {
-        int error;
-        cl_device_fp_config c;
-        error = clGetDeviceInfo(device, CL_DEVICE_DOUBLE_FP_CONFIG, sizeof(c),
-                                (void *)&c, NULL);
-        if (error)
+        if (lhs.data.s[i] != rhs.data.s[i])
         {
-            log_info(
-                "clGetDeviceInfo failed with CL_DEVICE_DOUBLE_FP_CONFIG\n");
             return false;
         }
-        return c != 0;
     }
-};
+    return true;
+}
+
+// compare for half vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
+    {
+        if (lhs.data.s[i] != rhs.data.s[i])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// compare for scalars
+template <typename Ty>
+typename std::enable_if<std::is_scalar<Ty>::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
+
+// compare for scalar halfs
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs.data == rhs.data;
+}
+
+template <typename Ty> inline bool compare_ordered(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
 
+template <>
+inline bool compare_ordered(const subgroups::cl_half &lhs,
+                            const subgroups::cl_half &rhs)
+{
+    return cl_half_to_float(lhs.data) == cl_half_to_float(rhs.data)
+        && !is_half_nan(lhs.data);
```

Comment:
I commented on this earlier but it dropped off. If lhs.data is NaN, the equality comparison will be false per IEEE-754 rules and the following NaN check irrelevant.
If the intent is to follow those rules it's harmless but confusing. A comment would be helpful to state the intended behavior.

Reply:
removed